src/legacy/core_plugins/kibana/public/discover/np_ready/angular/discover_state.ts

```diff
@@ -0,0 +1,238 @@
+  /**
+   * Return the Appstate before the current app state, useful for diffing changes
+   */
+  getPreviousAppState: () => AppState;
+  /**
+   * Returns whether the current app state is different to the initial state
+   */
+  isAppStateDirty: () => void;
+}
+const APP_STATE_URL_KEY = '_a';
+
+/**
+ * Builds and returns appState and globalState containers and helper functions
+ * Used to sync URL with UI state
+ */
+export function getState({
+  defaultAppState = {},
+  storeInSessionStorage = false,
+  hashHistory,
+}: GetStateParams): GetStateReturn {
+  const stateStorage = createKbnUrlStateStorage({
+    useHash: storeInSessionStorage,
+    history: hashHistory ? hashHistory : createHashHistory(),
+  });
+
+  const appStateFromUrl = stateStorage.get(APP_STATE_URL_KEY) as AppState;
+  let initialAppState = {
+    ...defaultAppState,
+    ...appStateFromUrl,
+  };
+  let previousAppState: AppState;
+  const appStateContainer = createStateContainer<AppState>(initialAppState);
+
+  const appStateContainerModified = {
+    ...appStateContainer,
+    set: (value: AppState | null) => {
+      if (value) {
+        previousAppState = appStateContainer.getState();
+        appStateContainer.set(value);
+      }
+    },
+  };
+
+  const { start, stop } = syncState({
+    storageKey: APP_STATE_URL_KEY,
+    stateContainer: appStateContainerModified,
+    stateStorage,
+  });
+
+  return {
+    kbnUrlStateStorage: stateStorage,
+    appStateContainer: appStateContainerModified,
+    startSync: start,
+    stopSync: stop,
+    setAppState: (newPartial: AppState) => setState(appStateContainerModified, newPartial),
+    replaceUrlAppState: async (newPartial: AppState = {}) => {
+      const state = { ...appStateContainer.getState(), ...newPartial };
+      await stateStorage.set(APP_STATE_URL_KEY, state, { replace: true });
+    },
+    getAppFilters: () => getFilters(appStateContainer.getState()),
+    resetInitialAppState: () => {
+      initialAppState = appStateContainer.getState();
+    },
+    getPreviousAppState: () => previousAppState,
+    flushToUrl: () => stateStorage.flush(),
+    isAppStateDirty: () => !isEqualState(initialAppState, appStateContainer.getState()),
+  };
+}
+
+/**
+ * Helper function to merge a given new state with the existing state and to set the given state
+ * container
+ */
+export function setState(stateContainer: ReduxLikeStateContainer<AppState>, newState: AppState) {
+  const oldState = stateContainer.getState();
+  const mergedState = { ...oldState, ...newState };
+  if (!isEqualState(oldState, mergedState)) {
+    if (mergedState.query) {
+      mergedState.query = migrateLegacyQuery(mergedState.query);
+    }
+    stateContainer.set(mergedState);
+  }
+}
+
+/**
+ * Helper function to compare 2 different filter states
+ */
+export function isEqualFilters(filtersA: Filter[], filtersB: Filter[]) {
+  if (!filtersA && !filtersB) {
+    return true;
+  } else if (!filtersA || !filtersB) {
```

Comment:
Don't understand for which cases we need it: 
```
else if (!filtersA || !filtersB) {
    return false;
  }
```
I think `esFilters.compareFilters` should correctly handle that case

Reply:
what about this case? throws an exception. `compareFilters` just checks for `Array.isArray` 
```Typescript
 test('compareFilters works', async () => {
    const result = esFilters.compareFilters(null, {}, esFilters.COMPARE_ALL_OPTIONS)
    expect(result).toBeFalsy();
  });
```

![Bildschirmfoto 2020-03-05 um 16 22 20](https://user-images.githubusercontent.com/463851/75996029-92273000-5efd-11ea-967c-4edf11a07a80.png)

Reply:
Oh.... good reason to handle that case in esFilters.compareFilters method =)

Reply:
Yeah, this should be handled inside `compareFilters`

Reply:
I've added an issue  #59526 for that, just to be sure that the comment length of this PR won't exceed e.g. odyssey one day ;)